aiohttp/client.py

```diff
@@ -187,6 +217,44 @@ class ClientTimeout:
 class ClientSession:
     """First-class interface for making HTTP requests."""
 
+    if sys.version_info >= (3, 11):
```

Comment:
Just to double check, you've tested all these methods in mypy?

From my reading of the documentation, I'd expect this to be interpreted as 5 overloads of `__slots__` or something weird like that...

> A series of @overload-decorated definitions must be followed by exactly one non-@overload-decorated definition (for the same function/method).

Reply:
This works with Pyright, I can check MyPy a well though. Typically yes, you would have multiple overloads, but this a bit of a typing hack and a single overload should still get recognized.

Reply:
Okay, instead of overloading an `if TYPE_CHECKING` block ended up likely making more sense, and is less of a hack. Testing with MyPy:

Reply:
```python
import asyncio
import aiohttp
from typing_extensions import reveal_type


async def main() -> None:
    async with aiohttp.ClientSession() as session:
        async with session.get("https://api.github.com/events") as response:
            reveal_type(session.get)
            print("Status:", response.status)
            print("Content-type:", response.headers["content-type"])

            html = await response.text()
            print("Body:", html)


if __name__ == "__main__":
    asyncio.run(main())
```

We get:

>> aiohttp/test.py:9:25: note: Revealed type is "def (url: Union[builtins.str, yarl.URL], **kwargs: Unpack[TypedDict('aiohttp.client._RequestOptions', {'params'?: Union[typing.Mapping[builtins.str, builtins.str], None], 'data'?: Any, 'json'?: Any, 'cookies'?: Union[Union[typing.Mapping[builtins.str, Union[builtins.str, http.cookies.BaseCookie[builtins.str], http.cookies.Morsel[Any]]], typing.Iterable[tuple[builtins.str, Union[builtins.str, http.cookies.BaseCookie[builtins.str], http.cookies.Morsel[Any]]]], http.cookies.BaseCookie[builtins.str]], None], 'headers'?: Union[Union[typing.Mapping[Union[builtins.str, multidict.istr], builtins.str], multidict.CIMultiDict[builtins.str], multidict.CIMultiDictProxy[builtins.str]], None], 'skip_auto_headers'?: Union[typing.Iterable[builtins.str], None], 'auth'?: Union[tuple[Any, Any, Any, fallback=aiohttp.helpers.BasicAuth], None], 'allow_redirects'?: builtins.bool, 'max_redirects'?: builtins.int, 'compress'?: Union[builtins.str, None], 'chunked'?: Union[builtins.bool, None], 'expect100'?: builtins.bool, 'raise_for_status'?: Union[None, builtins.bool, def (aiohttp.client_reqrep.ClientResponse) -> typing.Awaitable[None]], 'read_until_eof'?: builtins.bool, 'proxy'?: Union[Union[builtins.str, yarl.URL], None], 'proxy_auth'?: Union[tuple[Any, Any, Any, fallback=aiohttp.helpers.BasicAuth], None], 'timeout'?: Union[aiohttp.client.ClientTimeout, aiohttp.helpers._SENTINEL, None], 'ssl'?: Union[ssl.SSLContext, builtins.bool, aiohttp.client_reqrep.Fingerprint], 'server_hostname'?: Union[builtins.str, None], 'proxy_headers'?: Union[Union[typing.Mapping[Union[builtins.str, multidict.istr], builtins.str], multidict.CIMultiDict[builtins.str], multidict.CIMultiDictProxy[builtins.str]], None], 'trace_request_ctx'?: Union[types.SimpleNamespace, None], 'read_bufsize'?: Union[builtins.int, None], 'auto_decompress'?: Union[builtins.bool, None], 'max_line_size'?: Union[builtins.int, None], 'max_field_size'?: Union[builtins.int, None]})]) -> aiohttp.client._RequestContextManager"
Success: no issues found in 1 source file

Which is what we would expect.

Reply:
> This works with Pyright, I can check MyPy as well though. Typically yes, you would have multiple overloads, but this a bit of a typing hack and a single overload should still get recognized.

I wasn't questioning the number of overloads, but their position. You seem to have assumed that an overload is based on the function name, but I can't find anything in the documentation or PEP that suggests the name of the function has any relevance for overloads. As per the quote from the documentation, the overload(s) would be followed by the implementation function.

So, I was expecting multiple separate overloads:
```
if sys.version_info ...:
    @overload
    async def get(...):
        ...
async get(...):
    [...]

if sys.version_info ...:
    @overload
    async def post(...):
        ....
async def post(...):
```

Reply:
Ah I see what you're saying, nonetheless this is no longer necessary with how we're doing this.

Reply:
Weirdly, when testing a small example in mypy I get errors that there must be multiple overloads and that the name already exists. Not sure why these same errors didn't occur in the PR...